File: packages/rules/src/dates.ts
/** Rules that check the year against reality (the `DT` family). */

import type { Diagnostic } from "@recite/core";
import { coverageLabel, editionsCoveringYear, findReporter } from "@recite/core";

import type { Rule, RuleContext } from "./rule.js";
import { diagnose } from "./rule.js";

/** Before this, nothing in the reporter tables exists; a lower year is a typo. */
const EARLIEST_PLAUSIBLE_YEAR = 1600;

/**
 * `999 F.3d 1 (2d Cir. 1950)` — the Federal Reporter's third series began in
 * 1993.
 *
 * The strongest check ReCite can perform without consulting anything: a
 * reporter series and a date are together enough to prove a citation
 * impossible, with no database and no network.
 */
export const yearOutsideReporterRange: Rule = {
  id: "DT001",
  name: "year-outside-reporter-range",
  summary: "Decision year falls outside the years that reporter was published.",
  severity: "error",

  check(ctx: RuleContext): Diagnostic[] {
    const found: Diagnostic[] = [];

    for (const citation of ctx.extraction.citations) {
      const canonical = citation.reporterCanonical;
      if (!canonical || citation.year === undefined) continue;

      const edition = findReporter(canonical);
      if (!edition) continue;
      if (
        citation.year >= edition.start &&
        (edition.end === null || citation.year <= edition.end)
      ) {
        continue;
      }

      const detail = `${canonical} was published ${coverageLabel(edition)}, but this cites ${citation.year}.`;
      const alternatives = editionsCoveringYear(canonical, citation.year);
      const only = alternatives.length === 1 ? alternatives[0] : undefined;

      if (only) {
        // Often a neighbouring edition of the same series was meant.
        found.push(
          diagnose(
            yearOutsideReporterRange,
            citation,
            `${detail} A ${citation.year} case in this series would be in ${JSON.stringify(only.abbrev)}.`,
            {
              replacement: citation.text.replace(
                citation.reporter ?? canonical,
                only.abbrev,
              ),
              safety: "unsafe",
              fixDescription: `Change the reporter to ${JSON.stringify(only.abbrev)}`,
              context: {
                reporter: canonical,
                year: citation.year,
                suggestion: only.abbrev,
              },
            },
          ),
        );
      } else {
        found.push(
          diagnose(
            yearOutsideReporterRange,
            citation,
            `${detail} Check the volume, the reporter and the year.`,
            { context: { reporter: canonical, year: citation.year } },
          ),
        );
      }
    }

    return found;
  },
};

/** A year that cannot be right on its face — in the future, or medieval. */
export const implausibleYear: Rule = {
  id: "DT002",
  name: "implausible-year",
  summary: "Decision year is in the future or impossibly early.",
  severity: "error",

  check(ctx: RuleContext): Diagnostic[] {
    const found: Diagnostic[] = [];

    for (const citation of ctx.extraction.citations) {
      if (citation.year === undefined) continue;

      if (citation.year > ctx.currentYear) {
        found.push(
          diagnose(
            implausibleYear,
            citation,
            `Cites a decision dated ${citation.year}, which is in the future (this year is ${ctx.currentYear}).`,
            { context: { year: citation.year } },
          ),
        );
      } else if (citation.year < EARLIEST_PLAUSIBLE_YEAR) {
        found.push(
          diagnose(
            implausibleYear,
            citation,
            `Decision year ${citation.year} predates any reporter in the tables — likely a typo.`,
            { severity: "warning", context: { year: citation.year } },
          ),
        );
      }
    }

    return found;
  },
};

File: packages/recite-rules/src/recite/rules/dates.py
"""Rules that check the year against reality (the ``DT`` family)."""

from __future__ import annotations

from collections.abc import Iterable

from recite.core import (
    Diagnostic,
    FixSafety,
    Severity,
    edition_info,
    editions_covering_year,
)

from .base import Rule, RuleContext

__all__ = ["ImplausibleYear", "YearOutsideEdition"]

#: Before this, nothing in the reporter databases exists; a year below it is a
#: typo rather than a very old case.
_EARLIEST_PLAUSIBLE_YEAR = 1600


class YearOutsideEdition(Rule):
    """``999 F.3d 1 (2d Cir. 1950)`` — F.3d did not exist until 1993.

    This is the single most useful offline check: it catches a fabricated or
    transposed citation without needing any network access, because a reporter
    series and a date are enough to prove the pairing impossible.
    """

    id = "DT001"
    name = "year-outside-edition"
    summary = "Decision year falls outside the years that reporter was published."
    severity = Severity.ERROR

    def check(self, ctx: RuleContext) -> Iterable[Diagnostic]:
        for citation in ctx.citations:
            reporter = citation.reporter_canonical
            if not reporter or citation.year is None:
                continue

            edition = edition_info(reporter)
            if edition is None or edition.covers(citation.year):
                continue

            # Often the author meant a neighbouring edition of the same series
            # — F.2d rather than F.3d — which we can name precisely.
            alternatives = editions_covering_year(reporter, citation.year)
            detail = (
                f"{reporter} was published {edition.coverage_label()}, "
                f"but this cites {citation.year}."
            )

            if len(alternatives) == 1:
                replacement = alternatives[0].name
                yield self.diagnostic(
                    citation,
                    f"{detail} A {citation.year} case in this series would be "
                    f"in {replacement!r}.",
                    replacement=citation.corrected.replace(reporter, replacement, 1),
                    safety=FixSafety.UNSAFE,
                    fix_description=f"Change reporter to {replacement!r}",
                    reporter=reporter,
                    year=citation.year,
                    suggestion=replacement,
                )
            else:
                yield self.diagnostic(
                    citation,
                    f"{detail} Check the volume, the reporter and the year.",
                    reporter=reporter,
                    year=citation.year,
                    coverage=edition.coverage_label(),
                )


class ImplausibleYear(Rule):
    """A year that cannot be right on its face — in the future, or medieval."""

    id = "DT002"
    name = "implausible-year"
    summary = "Decision year is in the future or impossibly early."
    severity = Severity.ERROR

    def check(self, ctx: RuleContext) -> Iterable[Diagnostic]:
        for citation in ctx.citations:
            if citation.year is None:
                continue

            if citation.year > ctx.current_year:
                yield self.diagnostic(
                    citation,
                    f"Cites a decision dated {citation.year}, which is in the "
                    f"future (this year is {ctx.current_year}).",
                    year=citation.year,
                )
            elif citation.year < _EARLIEST_PLAUSIBLE_YEAR:
                yield self.diagnostic(
                    citation,
                    f"Decision year {citation.year} predates any reporter in "
                    f"the database — likely a typo.",
                    severity=Severity.WARNING,
                    year=citation.year,
                )
